### DISCUSSION TITLE
Creating own library for a lithium ion cell

### DISCUSSION THREAD

**Discussion by mlnsai** [pybamm-team/PyBaMM#2452] [General]
Hi Team,

I can see inbuilt cell libraries such as  LGM50 (Chen2020), Enertech Cell (Ai2020) etc.

Can i create such library for a cell of my choice? (Let's say i have the manufacturing facility and i can measure/find out any parameter)

So, if i want to create a similar library for a new battery and run simulations on that. What all parameters do i require apart from "parameters" in "input" folder? 

I have gone through all the parameters the code takes from "cell", "electrolyte", "positive electrode", "negative electrode", "sei", "separator", "lithium plating".

Apart from these do i require any additional parameters to make my own library of a cell?

Is it possible at all? 

Thanks in advance.

**Comment by brosaplanella** [comment]
We recently reformatted the parameter structure, so this solution applies to PyBaMM 22.10 onwards. You can define your parameters directly as a dictionary, as shown in [this notebook](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/parameterization/parameterization.ipynb) (section `Using pre-defined models in PyBaMM`). Note that you can also define them as a separate file (following one of the examples [here](https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/input/parameters/lithium_ion/Chen2020.py)) and then import them in your main script.

**Reply by PaulKranefeld** [reply]
Hello, this raises the question for me, how can I import a custom parameter set (ownParameterset) that I have created as a separate Python file.(As in the Chen2020 example) The easiest would be to have that file in my working directory. I would like to call the parameter set via `pybamm.ParameterValues("ownParameterset")` How is that possible? I checked the related Notebooks but since the parameter management got updated they are a bit out of date. Thanks in advance.

**Reply by valentinsulzer** [reply]
@awadell1 can you help?

**Reply by awadell1** [reply]
Have you looked at [Adding a parameter set](https://pybamm.readthedocs.io/en/latest/source/parameters/parameter_sets.html#adding-parameter-sets)? The docs run through an example using `cell_alpha.py` substitute `ownParameterset` for `cell_alpha` and you'll be off to the races.

tldr; To be recognized as a named parameter set, you need to create a python package, then register an entry point to "pybamm_parameter_sets".

**Reply by PaulKranefeld** [reply]
Thank you for this quick feedback, I will try this way.

**Reply by PaulKranefeld** [reply]
Sorry to bother you again but I tried to follow the documentation and Pybamm does not recognize my parameter set yet. I have performed the following steps:
1. Setting up a package in my current working directory as it is explained in the docs (as i understood)
```
 working_directory
  ├──pybamm_simulation.py   
  └──cell_parameters
       ├── setup.py  
       ├──_init_.py      
       └── src
            ├──_init_.py
            └── cell_parameters
                 └── parameters_test.py
                 └── _init_.py
```
2. Adding content to setup.py file to create the entry point
```
from setuptools import setup

setup(
    # ...,
    entry_points = {
        "pybamm_parameter_sets": [
            "parameters_test = cell_parameters.parameters_test:get_parameter_values"
        ]
    }
)
```
3. In my simulation file i did the following:
````
import pybamm
from cell_parameters.src.cell_parameters import parameters_test

model = pybamm.lithium_ion.DFN(options, name="test", build=False)
params = pybamm.ParameterValues("parameters_test")
````
Where did I make a mistake?

**Reply by awadell1** [reply]
Did you install the package? From within `working_directory,` run `python3 -m pip install .`?

You do need to reinstall the package if you add new parameter sets. If you're editing your parameter set, you may want to install in edit mode: `python3 -m pip install -e .`

**Reply by PaulKranefeld** [reply]
Thank you for your help. I have not installed the package before and did that now. That worked when i executed `pip install .` in the `cell_parameters` directory but not in the `working_directory` (error here was: `ERROR: Directory '.' is not installable. Neither 'setup.py' nor 'pyproject.toml' found.`. When i run the `pybamm_simulation.py` now i get the following error: `cannot import name 'parameters_test' from 'cell_parameters'`

Edit: Is it possible that i did a mistake within the structure of the package? Do I have to place the setup.py file in the working directory and outside of the package?

**Reply by PaulKranefeld** [reply]
I now succesfully installed the package `cell_parameters` with `parameters_test` in it and it appears in site_packages within my virtual environement. Pybamm still does not recognize it if i do this:
```
import pybamm
import cell_parameters

model = pybamm.lithium_ion.DFN(options, name="Test", build=False)
params = pybamm.ParameterValues("parameters_test")
```
`FileNotFoundError: Could not find parameter parameters_test.`
Can someone help me?

**Reply by awadell1** [reply]
It should be `pybamm_parameter_set`, not `pybamm_parameter_sets` 

```python
from setuptools import setup

setup(
    # ...,
    entry_points = {
        "pybamm_parameter_set": [
            "parameters_test = cell_parameters.parameters_test:get_parameter_values"
        ]
    }
)
```

🤦‍♂️ Apologies for the issue, fixing the docs now

**Reply by PaulKranefeld** [reply]
Now it works like a charm! Thank you for your help.

**Reply by alexisfortini** [reply]
Hello, I was wondering if I could get some help with adding a parameter set. I have reviewed the documentation and this post but I am still having trouble.

These are the steps I took:

1.  Set up a package in my current working directory (where parameters_test.py is a copy of Chen2020.py)
```
working_directory
  ├──pybamm_simulation.py   
  └──cell_parameters
       ├── setup.py  
       └── src
            └── cell_parameters
                 └── parameters_test.py
```

2. Added content to setup.py file to create the entry point
```
from setuptools import setup

setup(
    # ...,
    entry_points = {
        "pybamm_parameter_set": [
            "parameters_test = cell_parameters.parameters_test:get_parameter_values"
        ]
    }
)
```

3. Ran ```pip install -e .``` from working_directory/cell_parameters. Now I can see a new folder ```cell_parameters.egg-info``` has been created in working_directory/src

4. Ran the ```pybamm_simulation.py``` simulation file. Line 2 successfully added the ```__pycache__``` folder to working_directory/src/cell_parameters, but I get ```FileNotFoundError: Could not find parameter parameters_test. If you have a developer install, try re-installing pybamm (e.g. `pip install -e .`) to expose recently-added parameter entry points.``` from line 5
```
import pybamm
from cell_parameters.src.cell_parameters import parameters_test

model = pybamm.lithium_ion.DFN(options, name="test", build=False)
params = pybamm.ParameterValues("parameters_test")
```

Where did I make a mistake? Thanks in advance for your help!

**Comment by valentinsulzer** [comment]
Try "pybamm_parameter_sets" as the entry point instead of "pybamm_parameter_set"

**Reply by alexisfortini** [reply]
Thanks for your quick response. I was able to successfully add my own parameter set with a few changes to my code. Here are the steps I took in case anyone else needs help: 

1.  Set up a package in my current working directory (where parameters_test.py is a copy of Chen2020.py)
```
working_directory
  ├──pybamm_simulation.py
  ├──setup.py   
  └──cell_parameters
       └── parameters_test.py
```

2. Added content to setup.py file to create the entry point
```
from setuptools import setup

setup(
    # ...,
    entry_points = {
        "pybamm_parameter_sets": [
            "parameters_test = cell_parameters.parameters_test:get_parameter_values"
        ]
    }
)
```

3. Ran ```pip install -e .``` from working_directory. Now I can see a new folder ```cell_parameters.egg-info``` has been created in working_directory

4. Ran the ```pybamm_simulation.py``` simulation file. Now I can see a new folder ```__pycache__``` has been created in working_directory/cell_parameters, ```parameters_test``` is now included in ```pybamm.parameter_sets```, and the model runs successfully.
```
import pybamm
from cell_parameters import parameters_test

print(list(pybamm.parameter_sets))

model = pybamm.lithium_ion.DFN()
chemistry = pybamm.parameter_sets.parameters_test
params = pybamm.ParameterValues(chemistry)
sim = pybamm.Simulation(model)
sim.solve([0, 3600])
sim.plot()
```

Thank you again for your help Dr. Sulzer!

**Reply by mlnsai** [reply]
Hey @alexisfortini and @valentinsulzer  , I am trying to follow your steps but I am unable to find any setup.py file to add my own entry point. Can you help me in adding my own parameter set? Here are the steps i followed.

1. Copied the "Chen2020.py" file contents to "Test.py" within the "\pybamm\input\parameters\lithium_ion" directory within my system.
2. Updated "__all__ = ['Ai2020', 'Chen2020', 'Chen2020_composite', 'Ecker2015',
           'Ecker2015_graphite_halfcell',  'MSMR_example_set', 'Marquis2019',
           'Mohtat2020', 'NCA_Kim2011', 'OKane2022',
           'OKane2022_graphite_SiOx_halfcell', 'ORegan2022', 'Prada2013',
           'Ramadass2004', 'Test', 'Xu2019']
" code in __init.py__ file within "\pybamm\input\parameters\lithium_ion" same path.
3. Ran sample code to get the error "    raise ValueError(
ValueError: 'Test' is not a valid parameter set. Parameter set must be one of:
Ai2020
Chayambuka2022
Chen2020
Chen2020_composite
ECM_Example
Ecker2015
Ecker2015_graphite_halfcell
MSMR_Example
Marquis2019
Mohtat2020
NCA_Kim2011
OKane2022
OKane2022_graphite_SiOx_halfcell
ORegan2022
Prada2013
Ramadass2004
Sulzer2019
Xu2019"

Can you please help me in adding the Test.py parameter set? I am using 25.1.1 version of Pybamm. Thanks in advance

@valentinsulzer @alexisfortini
